walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,113 @@
+    List<TimeRange> optionalRanges = determineAvailability(events, optionalMeetingAttendees, meetingDuration);
+    
+    // Return availability based upon the quantity of mandatory verus optional attendees 
+    if (mandatoryMeetingAttendees.isEmpty() && optionalMeetingAttendees.isEmpty()) {
+        // Return availability as the entire day if there are no mandatory nor optional attendees 
+        return Arrays.asList(TimeRange.WHOLE_DAY);
+    } else if (optionalMeetingAttendees.isEmpty()) {
+        // Return availability of the mandatory attendees if there are no optional attendees 
+        return mandatoryRanges; 
+    } else if (mandatoryMeetingAttendees.isEmpty()) {
+        // Return availability of the optional attendees if there are no mandatory attendees 
+        return optionalRanges; 
+    } else {
+        // Merge mandatory and optional attendees 
+        List<String> mergedAttendees = new ArrayList<>(mandatoryMeetingAttendees);
+        mergedAttendees.addAll(optionalMeetingAttendees);
+
+        // Return availability of all attendees if possible, otherwise return availability of mandatory attendees
+        List<TimeRange> mergedRanges = determineAvailability(events, mergedAttendees, meetingDuration);
+        return mergedRanges.isEmpty() ? mandatoryRanges : mergedRanges; 
+    }
+  }
+
+  // Determine the time ranges for which the meeting attendees can all mutually meet
+  private List<TimeRange> determineAvailability(Collection<Event> events, 
+    Collection<String> meetingAttendees, long meetingDuration) {
+
+    List<TimeRange> busyTimes = new ArrayList<>();
+
+    // Acquire the time range of every event that shares an attendee with the meeting 
+    for (Event event: events) {
+      Collection<String> eventAttendees = event.getAttendees();
+      if (hasMutualAttendees(eventAttendees, meetingAttendees)) {
+        busyTimes.add(event.getWhen()); 
+      }
+    }
+
+    // Sort these 'busy times' in order of ascending start time 
+    Collections.sort(busyTimes, TimeRange.ORDER_BY_START);
+
+    // Obtain the inverse of the 'busy times' (i.e. aquire the 'free times')
+    List<TimeRange> freeTimes = inversify(busyTimes, meetingDuration);
+
+    // Sort and return the 'free times' in order of ascending start time 
+    Collections.sort(freeTimes, TimeRange.ORDER_BY_START);
+    return freeTimes;
+  }
+
+  // Return true if an event and a meeting share an attendee, false otherwise
+  private boolean hasMutualAttendees(Collection<String> eventAttendees, Collection<String> meetingAttendees) {
+    for (String eventAttendee: eventAttendees) {
+       if (meetingAttendees.contains(eventAttendee)) {
+           return true; 
+       }
+    }
+    return false;
+  } 
+
+  // Return the inverse of a schedule such that each returned range is an acceptable time to host a meeting
+  private List<TimeRange> inversify(Collection<TimeRange> busyTimes, long duration) {
```

Comment:
added; I wasn't totally familiar with pure before, but after reading more about it, I believe every function can be marked static because (1) they will always return the same output given the same input and (2) none of them modify the input (as everything is read-only). Please let me know if I mis-interpreted this definition

Reply:
A "pure" function fits the mathematical definition of a function like you learn in algebra f(x) -> y. The function takes an input and gives an output with no side effects.

Usually a function is safe to mark static when it doesn't depend on a `this` instance.

Additionally, a function _must_ be static if it depends on a `static` variable, though I think static variables are a code smell and you should be wary of them.